packages/paste-core/components/popover/src/PopoverContainer.tsx

```diff
@@ -1,17 +1,34 @@
 import * as React from 'react';
-import {
-  useNonModalDialogPrimitiveState,
+import {Box} from '@twilio-paste/box';
+import type {
+  NonModalDialogPrimitiveStateReturn,
   NonModalDialogPrimitivePopoverInitialState,
 } from '@twilio-paste/non-modal-dialog-primitive';
+import {useNonModalDialogPrimitiveState} from '@twilio-paste/non-modal-dialog-primitive';
 import {PopoverContext} from './PopoverContext';
 
-export type PopoverContainerProps = NonModalDialogPrimitivePopoverInitialState;
+export interface PopoverStateReturn extends NonModalDialogPrimitiveStateReturn {
+  [key: string]: any;
+}
 
-const PopoverContainer: React.FC<PopoverContainerProps> = ({children, placement, gutter, ...initialState}) => {
-  const popover = useNonModalDialogPrimitiveState({placement, gutter, modal: true, ...initialState});
-  const value = React.useMemo(() => popover, Object.values(popover));
-  return <PopoverContext.Provider value={value}>{children}</PopoverContext.Provider>;
-};
+export interface PopoverContainerProps extends NonModalDialogPrimitivePopoverInitialState {
+  children: NonNullable<React.ReactNode>;
+  state?: PopoverStateReturn;
+}
+
+const PopoverContainer = React.forwardRef<HTMLDivElement, PopoverContainerProps>(
+  ({children, gutter, placement, state, ...initialState}, ref) => {
+    const popover = state || useNonModalDialogPrimitiveState({gutter, modal: true, placement, ...initialState});
+    const value = React.useMemo(() => popover, Object.values(popover));
```

Comment:
❓ If `state` is passed would this memoized value re-compute on each re-render of the `PopoverContainer`? Asking because I'm guessing `state` is and object, and I think objects are re-instantiated on all re-renders. Would it make sense to instead memoize the container itself using `React.memo`?

Reply:
You're right. `state` is an `object`, so yeah memoizing the container makes sense. I'll make that change.